Overhaul net_mgmt to use bit fields

jukkar opened the pull request:
The current net_mgmt API uses 32 bit enum values for network event codes. This means that user needs to check what event is being triggered because the individual bits do not tell that. This has led to great confusion over the years and is being discussed more in #88534 

This PR proposes that each network event is mapped to one specific bit in the layer. To make sure that there are enough bits available, the net_mgmt API is changed to use 64 bit event number values.

jukkar committed:
net: mgmt: Convert the mgmt API to use 64-bit masks

Instead of using 32 bit enum values for event numbers, convert
the code to use 64 bit long bit fields. This means that the
user API is changed to use 64 bit event values instead of 32
bit event values.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

JordanYates commented:
What changed from your https://github.com/zephyrproject-rtos/zephyr/pull/88495#issuecomment-2796903007 that 64 bits will eventually be insufficient?

JordanYates commented on include/zephyr/net/net_mgmt.h:
```diff
@@ -76,6 +76,29 @@ struct net_if;
 
 /** @endcond */
 
+/** Central place the definition of the layer codes (7 bit value) */
```
Defining these layer codes as integers rather than individual bits essentially makes the `global_event_mask` optimisation in `net_mgmt.c` pointless. This is no different from the current behaviour, but worth looking at if we are redoing the definitions of events in order to get it right the first time?

jukkar replied:
You mentioned earlier the problems with `net_mgmt_event_wait()` but it seems that I missed your point there as after looking the API it looks ok after all. Or were you just wanting to avoid checking enum values in the first place and just use bit values?

jukkar commented:
> What changed from your [#88495 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/88495#issuecomment-2796903007) that 64 bits will eventually be insufficient?

Nothing really, just wanted to experiment with this idea of having one bit / event thingy. I looked also using bitarray but that would be much bigger change. So just experimenting and path finding a solution for having one bit / event with minimal API changes.

jukkar commented:
From the current set of changes here, the first three are more or less generic and we could probably apply them regardless of what is being done with the actual event values.

rlubos commented on include/zephyr/net/net_mgmt.h:
```diff
@@ -36,38 +35,28 @@ struct net_if;
+#define NET_MGMT_EVENT_SYNCHRONOUS(mgmt_request) FIELD_GET(NET_MGMT_SYNC_EVENT_MASK, mgmt_request)
+#define NET_MGMT_GET_LAYER(mgmt_request)         FIELD_GET(NET_MGMT_LAYER_MASK, mgmt_request)
+#define NET_MGMT_GET_LAYER_CODE(mgmt_request)    FIELD_GET(NET_MGMT_LAYER_CODE_MASK, mgmt_request)
+#define NET_MGMT_GET_COMMAND(mgmt_request)       FIELD_GET(NET_MGMT_COMMAND_MASK, mgmt_request)
```
Will this return a bitmask? I'm concerned there would be some confusion about this, event codes are still expressed by integers (`NET_EVENT_IF_CMD_DOWN` etc), but inside the event itself (`NET_EVENT_IF_DOWN`) it's converted to bitmask. I saw that for example `conn_mgr` uses `NET_MGMT_GET_COMMAND()` and compares it directly with event code (https://github.com/zephyrproject-rtos/zephyr/blob/c9f149cf6f445d7e39681ed0509e7acab5dee25b/subsys/net/conn_mgr/events_handler.c#L40-L41)

jukkar replied:
Note that I only changed `tests/net/mgmt` tests to pass (at least the default test), so there are probably lot of things to change in the code still if we want to go with these changes.

jukkar replied:
The reason why that conn_mgr example would not work is that `NET_MGMT_GET_COMMAND()` would return the bits only (not the layer), but the actual event `NET_EVENT_IF_CMD_DOWN` will also contain the layer value.

jukkar replied:
Fixed this issue in the latest version, for example the `NET_EVENT_IF_CMD_DOWN` is now a bit field so that the macro `NET_MGMT_GET_COMMAND()` can return it correctly.

rlubos reviewed (commented):
I just wonder if we should differentiate the actual event codes (integers defined in for example `enum net_event_if_cmd`) from the even representation in the 64-bit event, it seems it could cause some confustion.

jukkar commented:
> I just wonder if we should differentiate the actual event codes (integers defined in for example `enum net_event_if_cmd`) from the even representation in the 64-bit event, it seems it could cause some confustion.

Yeah, I thought about the same. It is just a bigger API change so did not go that route yet.

jukkar commented:
I moved first three commits to #89083 as they are pretty generic and could be used in main already.

jukkar committed:
doc: migration: 4.2: Add information about net_mgmt API changes

Add information how the net_mgmt request handler is changed.
The event number type is changed from uint32_t to uint64_t to allow
the event command to be a bit mask instead of enum value.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

pdgendt reviewed (dismissed):
This is trivial enough IMO, not a total rewrite needed.

jukkar commented:
* CI fixes

JordanYates reviewed (dismissed):
Direction is good, pending CI passing

jukkar committed:
net: socket: mgmt: Create proper socket options for net_mgmt sockets

We cannot use the network management event number directly as
a socket option value because the management value is uint64_t
and that cannot be mapped directly to 32 bit integer.
So have an intermediate socket option that is mapped to actual
network management request number in getsockopt() and setsockopt().

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* More CI fixes, this looks like a never ending saga as the net_mgmt API is used all over the place in zephyr
* I also noticed that net_mgmt API is used in hal_nxp which is used in CI for some tests and will need fixes

jukkar commented on include/zephyr/net/socket_net_mgmt.h:
```diff
@@ -42,6 +42,25 @@ extern "C" {
+/** Get number of priority queues */
+#define SO_NET_MGMT_ETHERNET_GET_PRIORITY_QUEUES_NUM 3
+
+/** Set Ethernet MAC address */
+#define SO_NET_MGMT_ETHERNET_SET_MAC_ADDRESS 4
```
I can rework the test so that we do not need to create non implemented socket options.

jukkar committed:
net: mgmt: Increment API major version number

As per API overview documentation, a braking API change must
increment major version number.

https://docs.zephyrproject.org/latest/develop/api/overview.html#api-overview

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* Rebased to latest main to fix merge conflict in manifest file
* Added info to release note
* Incremented net_mgmt API number as instructed by API lifecycle documentation https://docs.zephyrproject.org/latest/develop/api/overview.html#api-overview

decsny commented:
update manifest and remove DNM once #91494 merges due to NXP HAL synchronizing order
